File: screenshot/ViewControllers/BaseViewController.m
//
//  BaseViewController.m
//  screenshot
//
//  Created by Corey Werner on 8/8/17.
//  Copyright © 2017 crazeapp. All rights reserved.
//

#import "BaseViewController.h"
#import "screenshot-Swift.h"

@interface BaseViewController()

@property (nonatomic) BOOL isStatusBarHidden;

@end

@implementation BaseViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.view.backgroundColor = [UIColor background];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    if ([self.lifeCycleDelegate respondsToSelector:@selector(viewController:willAppear:)]) {
        [self.lifeCycleDelegate viewController:self willAppear:animated];
    }
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];

    if ([self.lifeCycleDelegate respondsToSelector:@selector(viewController:didAppear:)]) {
        [self.lifeCycleDelegate viewController:self didAppear:animated];
    }
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];

    if ([self.lifeCycleDelegate respondsToSelector:@selector(viewController:willDisappear:)]) {
        [self.lifeCycleDelegate viewController:self willDisappear:animated];
    }
}

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];

    if ([self.lifeCycleDelegate respondsToSelector:@selector(viewController:didDisappear:)]) {
        [self.lifeCycleDelegate viewController:self didDisappear:animated];
    }
}


#pragma mark - Status Bar

- (BOOL)prefersStatusBarHidden {
    return self.isStatusBarHidden;
}

- (void)showStatusBar {
    self.isStatusBarHidden = NO;
    [self setNeedsStatusBarAppearanceUpdate];
}

- (void)hideStatusBar {
    self.isStatusBarHidden = YES;
    [self setNeedsStatusBarAppearanceUpdate];
}


#pragma mark - Extra

- (void)addNavigationItemLogo {
    self.navigationItem.titleView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"Logo20h"]];
}

@end

File: screenshot/View/Generic/BaseViewController.swift
//
//  BaseViewController.swift
//  screenshot
//
//  Created by Gershon Kagan on 2/11/18.
//  Copyright © 2018 crazeapp. All rights reserved.
//

import UIKit

class BaseViewController: UIViewController {
    
    weak var lifeCycleDelegate: ViewControllerLifeCycle?
    var isStatusBarHidden = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = UIColor.background
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        lifeCycleDelegate?.viewController?(self, willAppear: animated)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        lifeCycleDelegate?.viewController?(self, didAppear: animated)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        lifeCycleDelegate?.viewController?(self, willDisappear: animated)
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        lifeCycleDelegate?.viewController?(self, didDisappear: animated)
    }
    
    // MARK: - Status Bar
    
    

    func showStatusBar() {
        self.isStatusBarHidden = false
        self.setNeedsStatusBarAppearanceUpdate()
    }
    func hideStatusBar() {
        self.isStatusBarHidden = true
        self.setNeedsStatusBarAppearanceUpdate()
    }
    
    override var prefersStatusBarHidden: Bool {
        return isStatusBarHidden
    }
    
    // MARK: - Extra
    
    @objc func addNavigationItemLogo() {
        self.navigationItem.titleView = UIImageView(image: UIImage(named: "Logo20h"))
    }
    
}

